improve-ephemeral-cluster-handling.md

```diff
@@ -0,0 +1,419 @@
+does not complete with success, we can imagine that it may be restarted
+even if technically the pod ran to completion. In that case we may want
+to allow an ephemeral cluster to remain
+
+### Add a CLI command to associate a driver with a long-running cluster
+
+Currently we only do association on create. However, the label mechanism
+will work to associate a driver with a long-running cluster, so we may
+want to add a CLI command to handle association/disassociation of
+drivers with long-runnign clusters.
+
+### Add a CLI command to lookup all the drivers using a shared cluster
+
+Since shared (long-running) clusters may be used by multiple drivers, given
+the above we may want to add a command to lookup all the drivers using
+a shared cluster. Or maybe it just prints as part of `get`.
+
+### Allow a driver to use more than one cluster
+
+Association currently is just a simple key/value pair. Not sure it makes
+sense for a driver to be associated with more than one cluster, but maybe.
+In that case it needs to be a list.
+
+### Allow clusters to be shared by multiple deployments from the same dc
+
+This really comes down to allowing an ephemeral cluster to persist across
+a redeploy of the driver. This is possible, but oshinko needs logic to
+look at the current state of the deploymentconfig and determine if it
+is spinning up a successor replicationcontroller while spinning down the
+current one.  If we can figure that out, this is possible.
+
+It begs the question of whether or not we always want to recreate clusters
+if we redeploy the driver, and this may be informed by a deeper look into
+how we do rolling upgrades (investigation coming soon).
+
+So maybe we don't want to pursue this yet ...
+
+### Move CLI extensions for ephemeral clusters to normal build
+
+The extra CLI command options are currently all in the `make extended`
+version of oshinko-cli and are only used in the S2I pods. We could consider
+moving those options to the normal CLI.
+
+However, we don't want to confuse users. If users are creating "ephemeral"
+clusters outside of an S2I workflow, maybe the user app that does the creating
+should be tracking clusters itself ...
+
+### Add a watcher container to the spark master pod to check for abandoned clusters
+
+This would be a simple container on the spark master pod that simply polls
+the oshinko-cli using some not-yet-invented command to check if the cluster
+is abandoned, i.e. it's ephemeral and the driver deployment cannot be located.
+
+The extra container could be added only to ephemeral clusters, and it could
+poll slowly (once every 5 seconds).
+
+If it found itself to be abandoned, it could call oshinko-cli delete on itself.
+This should work, because kubernetes sends a SIGTERM with a 30 second delay
+on a delete.  The call should complete.
+
+There may not be known edge cases where we end up with abandoned clusters
+at the moment, but this would guarantee cleanup if such edge cases or
+errors exist.
+
```

Comment:
from slack
> [9:36 AM] 
tmckay @elmiko, on the ephemeral cluster spec PR, "this seems tough" on the watcher container, what do you mean exactly?  Technically I think it's not actually hard

i just meant that this seems like there could be many edge cases where we need to work around specific conditions. i think on the surface it doesn't sound too challenging to monitor and setup the basic abandoned cluster reap, especially if we are doing it on a slow poll. but, i feel like there will be a bunch of corner cases that we might end up chasing to make this work nicely in production.

that's just my gut feeling though, i may be overestimating the difficulty